media-proxy/include/mesh/st2110tx.h

```diff
@@ -0,0 +1,153 @@
+#ifndef ST2110TX_H
+#define ST2110TX_H
+
+#include "st2110.h"
+#include <algorithm>
+
+namespace mesh::connection {
+
+/**
+ * ST2110
+ *
+ * Base abstract class of ST2110. ST2110_20Tx/ST2110_22Tx/ST2110_30Tx
+ * inherit this class.
+ */
+template <typename FRAME, typename HANDLE, typename OPS> class ST2110Tx : public ST2110 {
+  public:
+    ST2110Tx()
+    {
+        _kind = Kind::transmitter;
+
+        _handle = nullptr;
+        _ops = {0};
+        _transfer_size = 0;
+    };
+    ~ST2110Tx() { shutdown(_ctx); };
+
+  protected:
+    HANDLE _handle;
+    OPS _ops;
+    uint32_t _transfer_size;
+    context::Context _ctx;
+
+    virtual FRAME *get_frame(HANDLE) = 0;
+    virtual int put_frame(HANDLE, FRAME *) = 0;
+    virtual HANDLE create_session(mtl_handle, OPS *) = 0;
+    virtual int close_session(HANDLE) = 0;
+
+    Result on_establish(context::Context &ctx) override
+    {
+        _ctx = context::WithCancel(ctx);
+        _stop = false;
+
+        _handle = create_session(_st, &_ops);
+        if (!_handle) {
+            log::error("Failed to create session");
+            set_state(ctx, State::closed);
+            return set_result(Result::error_general_failure);
+        }
+
+        set_state(ctx, State::active);
+        return set_result(Result::success);
+    };
+
+    Result on_shutdown(context::Context &ctx) override
+    {
+        _ctx.cancel();
+
+        if (_handle) {
+            close_session(_handle);
+            _handle = nullptr;
+        }
+        set_state(ctx, State::closed);
+        return set_result(Result::success);
+    };
+
+    Result on_receive(context::Context &ctx, void *ptr, uint32_t sz, uint32_t &sent) override
+    {
+        sent = std::min(_transfer_size, sz);
+        // TODO: add error/warning if sent is different than _transfer_size
+
+        FRAME *frame = NULL;
+        do {
+            // Get empty buffer from MTL
+            frame = get_frame(_handle);
+            if (!frame) {
+                std::unique_lock lk(_mx);
+                _cv.wait(lk, _ctx.stop_token(), [this] { return _stop.load(); });
+                _stop = false;
+                if (_ctx.cancelled()) {
+                    break;
+                }
+            }
```

Comment:
Avoid using conditional variable and/or mutex in the hot path. Use atomics only.

Reply:
Do you suggesting to use spin locks with atomic variable instead of conditional variable? I'm opting against it. Sure, it would be faster, but it would consume a whole core while waiting. Here we want to retrieve the buffer from a ring, it is not that time-critical as it will be buffered anyway until it is sent.

Reply:
C++ introduced `.wait()`. Read this https://en.cppreference.com/w/cpp/atomic/atomic/wait

Reply:
This had been a very anticipated feature for long time.

Reply:
We already use it in the base connection class https://github.com/OpenVisualCloud/Media-Communications-Mesh/blob/3ade772ee4ae200c2186869215bf5e1ac5fca557/media-proxy/src/mesh/conn.cc#L201

Reply:
oh, cool. I didn't know about that one. In this case it is ok. 

I see that std::atomic::wait is even a little bit more readable.